pkg/model/combine_merging.go

```diff
@@ -0,0 +1,152 @@
+package model
+
+import (
+	"bytes"
+	"hash/fnv"
+
+	"github.com/pkg/errors"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/grafana/tempo/tempodb/encoding/common"
+)
+
+type ObjectMerger interface {
+	ID() common.ID
+	Merge(dataEncoding string, objs ...[]byte) (int, error)
+	Result(dataEncoding string) ([]byte, error)
+}
+
+type objectMerger struct {
+	lazyCurrent []byte
+	current     *tempopb.Trace
+	id          common.ID
+	spans       map[uint32]struct{}
+}
+
+// extractSpans creates a set which contains all spans that already exist in the merged structure
+func extractSpans(t *tempopb.Trace) map[uint32]struct{} {
+	m := make(map[uint32]struct{})
+	h := fnv.New32()
+	buffer := make([]byte, 4)
+	for _, batch := range t.Batches {
+		for _, ils := range batch.InstrumentationLibrarySpans {
+			for _, span := range ils.Spans {
+				m[tokenForID(h, buffer, int32(span.Kind), span.SpanId)] = struct{}{}
+			}
+		}
+	}
+	return m
+}
+
+func (m objectMerger) ID() common.ID {
+	return m.id
+}
+
+var _ = MergerFactory(NewMerger)
+
+type MergerFactory func(id []byte, initial []byte) ObjectMerger
+
+func NewMerger(id []byte, initialObj []byte) ObjectMerger {
+	return &objectMerger{
+		id:          id,
+		lazyCurrent: initialObj,
+	}
+}
+
+func (m *objectMerger) Merge(dataEncoding string, objs ...[]byte) (int, error) {
```

Comment:
AFAICT we don't have any cases that pass more than one `objs` slice. Should we just tighten up the interface and remove the loops?